### PR TITLE
LPS-185674 Add relationship, position, and linked features to model builder

### PR DESCRIPTION
**Pull request by igor-franca**
I performed a rebase with Brian's master to get the newly merged commits containing the ObjectFolderItem.

Therefore, the review of this PR can be started from here b84dfa3489bfbd3ec4aca1d94a6b57abcaa4583d.

Adds the features:

Relationship, positioning and linking of nodes.
Edit folder's ERC.

### CODE REVIEW

**Comment by naraandrade** [comment]
Depends on https://github.com/liferay-objects/liferay-portal/pull/2504

**Comment by naraandrade** [comment]
Just started reviewing :)

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/tests/FoldersListSidebar.tsx [hunk 1]
```diff
@@ -20,7 +28,7 @@ const uncategorizedFolder = {
 };
 
 const ticketFolder = {
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/tests/FoldersListSidebar.tsx:31]: @igor-franca Please, declare this const before uncategorizedFolder.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/index.d.ts [hunk 2]
```diff
@@ -308,14 +308,21 @@ interface ObjectFieldView extends ObjectField {
...
 	dateModified: string;
-	definitions?: ObjectDefinition[];
+	definitions?: ObjectDefinitionNodeData[];
 	externalReferenceCode: string;
 	id: number;
 	label: LocalizedValue<string>;
 	name: string;
+	objectFolderItems: ObjectFolderItem[];
+}
+interface ObjectFolderItem {
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/index.d.ts:321]: @igor-franca This interface needs to be declared before `ObjectFolder`.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/LeftSidebar/LeftSidebar.scss [hunk 3]
```diff
@@ -6,6 +6,11 @@
 	padding: 1rem;
 	position: inherit;
 
+	&-item-linked.treeview-link {
+		background-color: transparent;
+		color: rgba(107, 108, 126, 0.5);
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/LeftSidebar/LeftSidebar.scss:11]: @alinedoleron @igor-franca Could you please use the hexadecimal code for this color? We don't usually use rgba.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Header/Header.scss [hunk 4]
```diff
@@ -43,7 +54,15 @@
 			font-size: 18px;
 			font-weight: 600;
 			line-height: 150%;
+			max-width: 12.5rem;
+			overflow: hidden;
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Header/Header.scss:58]: @igor-franca `overflow` should be declared after `padding-right`.

**Review by naraandrade — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Header/Header.scss (hunk 4 above)
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Header/Header.scss:58]: hey @naraandrade, I think you got confused here, can you check it please?

Comment (naraandrade) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Header/Header.scss:58]: Oh sorry, my mistake!

**Comment by igor-franca** [comment]
Thanks for the review @naraandrade , I added a comment to confirm alphabetical ordering

**Review by naraandrade — APPROVED** [review]

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/EditObjectFolder.tsx [hunk 5]
```diff
@@ -35,18 +34,16 @@ export default function EditObjectFolder({
 		dispatch,
 	] = useFolderContext();
 
-	const [showModal, setShowModal] = useState<ViewObjectDefinitionsModals>({
+	const [showModal, setShowModal] = useState<ModelBuilderModals>({
 		addFolder: false,
 		addObjectDefinition: false,
-		bindToRootObjectDefinition: false,
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/EditObjectFolder.tsx]: @igor-franca @rebsilva is it safe to remove this property and the "unbind" one?

Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/EditObjectFolder.tsx]: These three properties belong to the Root Concept, in [creating](https://github.com/liferay-objects/liferay-portal/pull/2524/commits/590ef9aacd411d05f16e9896361c10a51c001600#diff-63608e745a4c720211693896469a7bf97d034829944b98edab0d2ed29a092d7cR53-R58) the ModelBuilderModals interface I omitted these three properties.

Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/EditObjectFolder.tsx]: The reason for bringing this commit to this PR was to not add Root Concept properties within ModelBuilder

Comment (naraandrade) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/EditObjectFolder.tsx]: Oh, I understand. Thanks for the explanation :)

**Comment by naraandrade** [comment]
ci:test:sf

**Comment by naraandrade** [comment]
ci:test:relevant

**Comment by igor-franca** [comment]
ci:test:object

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModalImportObjectDefinition.tsx [hunk 6]
```diff
@@ -69,7 +69,12 @@ export default function ModalImportObjectDefinition({
 
 	const handleImport = async (formData: FormData) => {
 		try {
-			await API.save(importObjectDefinitionURL, formData, 'POST');
+			await API.save({
+				item: formData,
+				method: 'POST',
+				returnValue: false,
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModalImportObjectDefinition.tsx:75]: @igor-franca The default returnValue is already false, I don't think we need to pass this property here, right?

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/LeftSidebar/LeftSidebar.tsx [hunk 7]
```diff
@@ -108,11 +108,11 @@ export default function LeftSidebar({
 			};
 
 			try {
-				const newObjectDefinition = (await API.save(
-					`/o/object-admin/v1.0/object-definitions/${objectDefinition?.id}`,
-					movedObjectDefinition,
-					'PATCH'
-				)) as ObjectDefinition;
+				const newObjectDefinition = (await API.save({
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/LeftSidebar/LeftSidebar.tsx:111]: @igor-franca In this case, shouldn't we be passing the property `returnValue: true`?

**Review by naraandrade — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModalImportObjectDefinition.tsx (hunk 6 above)
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModalImportObjectDefinition.tsx:75]: Well noted, thanks! 👍🏾

**Comment by igor-franca** [comment]
ci:test:sf

**Comment by igor-franca** [comment]
ci:test:object

**Review by naraandrade — APPROVED** [review]

**Comment by igor-franca** [comment]
ci:forward:force

**Comment by naraandrade** [comment]
Just started reviewing :)

**Pull request closed without merging**